lib/common.go

```diff
@@ -26,18 +29,120 @@ const (
 // the slave knows from which string permutation to start and at which string
 // permutation to stop.
 type Task struct {
-	ID        int
-	JobID     int
-	Algorithm string
-	Salt      string
-	Digest    string
-	CharSet   string
-	Length    int
-	Start     string
-	End       string
+	Job
+	JobID   int
+	ID      int
+	Start   []byte
+	TaskLen int
 }
 
 // ToJSON serializes a Task to JSON.
 func (t *Task) ToJSON() ([]byte, error) {
 	return json.Marshal(t)
 }
+
+type Algorithm int
+
+const (
+	PBKDF2 Algorithm = iota
+	BCRYPT
+	SCRYPT
+	ARGON2
+)
+
+type CharSet int
+
+const (
+	Numerical CharSet = iota
+	AlphaLower
+	AlphaMixed
+	AlphaNumLower
+	AlphaNumMixed
+)
+
+// Job is the customer facing resource representing a single password cracking job
+// we focus on PBKDF2 first with SHA256 first
+type Job struct {
+	Salt      []byte
+	Digest    []byte
+	KeyLen    int
+	Iter      int
+	CharSet   CharSet
+	Algorithm Algorithm
+}
+
+// CharSetSlice contains the set of all candidate characters for every alphabet
+var CharSetSlice [][]byte
+
+func init() {
+	const nums string = "0123456789"
+	const alphaLower string = "abcdefghijklmnopqrstuvwxyz"
+	const alphaMixed string = alphaLower + "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	const alphaNumLower string = nums + alphaLower
+	const alphaNumMixed string = nums + alphaMixed
+	CharSetSlice = [][]byte{
+		[]byte(nums),
+		[]byte(alphaLower),
+		[]byte(alphaMixed),
+		[]byte(alphaNumLower),
+		[]byte(alphaNumMixed)}
+}
+
+func BytesToIntSlice(charset CharSet, inp []byte) []int {
+	res := make([]int, len(inp))
+	for i, b := range inp {
+		// probably not efficient, but the character sets are small so it's negligible
+		x := bytes.IndexByte(CharSetSlice[charset], b)
+		if x < 0 {
+			panic("Invalid characters!")
+		}
+		res[i] = x
+	}
+	return res
+}
+
+func AddToIntSlice(base, v int, inp []int) ([]int, int) {
+	for i := range inp {
+		r := v % base
+		tmp := inp[i] + r
+		if tmp < base {
+			inp[i] = tmp
+			v = v / base
+		} else {
+			inp[i] = tmp % base
+			v = v/base + 1
+		}
+	}
+	return inp, v // v is the carry
+}
+
+func IntSliceToBytes(charset CharSet, inp []int) []byte {
+	res := make([]byte, len(inp))
+	for i, v := range inp {
+		res[i] = CharSetSlice[charset][v]
+	}
+	return res
+}
+
+func TestEqInts(a, b []int) bool {
```

Comment:
I'm keeping the `TestEqInts`. Using `reflect.DeepEqual` caused `job_test.go` to jump from 0.12 seconds to 0.5 seconds.